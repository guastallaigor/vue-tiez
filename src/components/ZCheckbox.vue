<template>
  <label class="z-checkbox">
    <span>
      <input type="checkbox" v-model="valueComp">
      <svg width='12' height='12' viewBox='0 0 12 12' fill='none' xmlns='http://www.w3.org/2000/svg' :class="{ disabled }" v-show="valueComp">
        <mask id='mask0' mask-type='alpha' maskUnits='userSpaceOnUse' x='0' y='0' width='12' height='12'>
          <path d='M9 0H3C1.34315 0 0 1.34315 0 3V9C0 10.6569 1.34315 12 3 12H9C10.6569 12 12 10.6569 12 9V3C12 1.34315 10.6569 0 9 0Z' fill='white'/>
        </mask>
        <g mask='url(%23mask0)'>
          <path d='M9 0H3C1.34315 0 0 1.34315 0 3V9C0 10.6569 1.34315 12 3 12H9C10.6569 12 12 10.6569 12 9V3C12 1.34315 10.6569 0 9 0Z' :fill='getFillChecked'/>
          <path d='M3.53906 4.79101L6.68945 7.88867L15.1213 -0.543182' :stroke='dark ? "black" : "white"' stroke-width='1.5'/>
        </g>
      </svg>
      <svg width='12' height='12' viewBox='0 0 12 12' fill='none' xmlns='http://www.w3.org/2000/svg' :class="{ disabled }" v-show="!valueComp">
        <path d='M8.5 0.5H3.5C1.84315 0.5 0.5 1.84315 0.5 3.5V8.5C0.5 10.1569 1.84315 11.5 3.5 11.5H8.5C10.1569 11.5 11.5 10.1569 11.5 8.5V3.5C11.5 1.84315 10.1569 0.5 8.5 0.5Z' :fill='getFill' stroke='#999'/>
      </svg>
    </span>
  </label>
</template>

<script>
export default {
  name: 'z-checkbox',
  props: {
    value: {
      type: Boolean,
      required: true,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    dark: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    valueComp: {
      get () {
        return this.value
      },
      set (val) {
        if (this.disabled) return

        this.$emit('input', val)
      }
    },
    getFill () {
      const { disabled, dark } = this

      if (!disabled) return ''

      return dark ? '#333' : '#EFEFEF'
    },
    getFillChecked () {
      const { disabled, dark } = this

      if (disabled) return '#999999'

      if (dark) return 'white'

      return 'black'
    }
  }
}
</script>

<style lang="scss">
.z-checkbox {
  cursor: default;

  span {
    display: inline-flex;
    position: relative;

    input[type="checkbox"] {
      margin: 3px 3px 3px 4px;
    }

    svg {
      cursor: pointer;
      display: inline-block;
      height: 12px;
      width: 12px;
      outline: 0;

      &.disabled {
        cursor: not-allowed;
      }
    }
  }
}
</style>
