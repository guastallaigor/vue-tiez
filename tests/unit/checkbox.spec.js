import { shallowMount } from '@vue/test-utils'
import ZCheckbox from '../../src/components/ZCheckbox'

describe('When I create the ZCheckbox component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZCheckbox, {
      propsData: propsData,
      slots: {
        default: slot || 'ZCheckbox'
      }
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent({ value: true })
    expect(wrapper.find(ZCheckbox).isVueInstance()).toBe(true)
  })

  it('should have a dark mode when unchecked', () => {
    const wrapper = createComponent({ value: false, dark: true })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const path = wrapper.find('.z-checkbox > span > svg > path')
    expect(path.exists()).toBe(true)
    expect(path.attributes().fill).toBe('')
  })

  it('should have a dark mode when unchecked and disabled', () => {
    const wrapper = createComponent({ value: false, dark: true, disabled: true })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const path = wrapper.find('.z-checkbox > span > svg > path')
    expect(path.exists()).toBe(true)
    expect(path.attributes().fill).toBe('#333')
  })

  it('should have a dark mode when checked', () => {
    const wrapper = createComponent({ value: true, dark: true })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const pathFirst = wrapper.find('.z-checkbox > span > svg > g > path:first-child')
    expect(pathFirst.exists()).toBe(true)
    expect(pathFirst.attributes().fill).toBe('white')
    const pathLast = wrapper.find('.z-checkbox > span > svg > g > path:last-child')
    expect(pathLast.exists()).toBe(true)
    expect(pathLast.attributes().stroke).toBe('black')
  })

  it('should have a dark mode when checked and disabled', () => {
    const wrapper = createComponent({ value: true, dark: true, disabled: true })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const pathFirst = wrapper.find('.z-checkbox > span > svg > g > path:first-child')
    expect(pathFirst.exists()).toBe(true)
    expect(pathFirst.attributes().fill).toBe('#999999')
    const pathLast = wrapper.find('.z-checkbox > span > svg > g > path:last-child')
    expect(pathLast.exists()).toBe(true)
    expect(pathLast.attributes().stroke).toBe('black')
  })

  it('should have a default state when unchecked', () => {
    const wrapper = createComponent({ value: false })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const path = wrapper.find('.z-checkbox > span > svg > path')
    expect(path.exists()).toBe(true)
    expect(path.attributes().fill).toBe('')
  })

  it('should have a default state when checked', () => {
    const wrapper = createComponent({ value: true })
    const checkbox = wrapper.find('.z-checkbox')
    expect(checkbox.exists()).toBe(true)
    const pathFirst = wrapper.find('.z-checkbox > span > svg > g > path:first-child')
    expect(pathFirst.exists()).toBe(true)
    expect(pathFirst.attributes().fill).toBe('black')
    const pathLast = wrapper.find('.z-checkbox > span > svg > g > path:last-child')
    expect(pathLast.exists()).toBe(true)
    expect(pathLast.attributes().stroke).toBe('white')
  })

  it('should have a disabled class when unchecked', async () => {
    const wrapper = createComponent({ value: false, disabled: true })
    const checkbox = wrapper.find('.z-checkbox > span > input')
    expect(checkbox.exists()).toBe(true)
    const svg = wrapper.find('.z-checkbox > span > svg')
    expect(svg.exists()).toBe(true)
    expect(svg.classes().length).toBe(1)
    expect(svg.classes()).toEqual(['disabled'])
    checkbox.setChecked()
    checkbox.trigger('input')
    wrapper.vm.valueComp = false
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().input).toBeFalsy()
  })

  it('should have a disabled class when checked', async () => {
    const wrapper = createComponent({ value: true, disabled: true })
    const checkbox = wrapper.find('.z-checkbox > span > input')
    expect(checkbox.exists()).toBe(true)
    const svg = wrapper.find('.z-checkbox > span > svg')
    expect(svg.exists()).toBe(true)
    expect(svg.classes().length).toBe(1)
    expect(svg.classes()).toEqual(['disabled'])
    checkbox.setChecked()
    checkbox.trigger('input')
    wrapper.vm.valueComp = false
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().input).toBeFalsy()
  })

  it('should be able to check when unchecked', async () => {
    const wrapper = createComponent({ value: false })
    const checkbox = wrapper.find('.z-checkbox > span > input')
    expect(checkbox.exists()).toBe(true)
    checkbox.setChecked()
    checkbox.trigger('input')
    wrapper.vm.valueComp = true
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().input).toBeTruthy()
  })

  it('should be able to uncheck when checked', async () => {
    const wrapper = createComponent({ value: true })
    const checkbox = wrapper.find('.z-checkbox > span > input')
    expect(checkbox.exists()).toBe(true)
    checkbox.setChecked()
    checkbox.trigger('input')
    wrapper.vm.valueComp = false
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted().input).toBeTruthy()
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      value: false,
      disabled: false,
      dark: false
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
