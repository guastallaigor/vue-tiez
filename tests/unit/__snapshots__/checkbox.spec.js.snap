// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZCheckbox component should match snapshot 1`] = `<label class="z-checkbox"><span><input type="checkbox"> <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg" class="" style="display: none;"><mask id="mask0" mask-type="alpha" maskUnits="userSpaceOnUse" x="0" y="0" width="12" height="12"><path d="M9 0H3C1.34315 0 0 1.34315 0 3V9C0 10.6569 1.34315 12 3 12H9C10.6569 12 12 10.6569 12 9V3C12 1.34315 10.6569 0 9 0Z" fill="white"></path></mask> <g mask="url(%23mask0)"><path d="M9 0H3C1.34315 0 0 1.34315 0 3V9C0 10.6569 1.34315 12 3 12H9C10.6569 12 12 10.6569 12 9V3C12 1.34315 10.6569 0 9 0Z" fill="black"></path> <path d="M3.53906 4.79101L6.68945 7.88867L15.1213 -0.543182" stroke="white" stroke-width="1.5"></path></g></svg> <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg" class=""><path d="M8.5 0.5H3.5C1.84315 0.5 0.5 1.84315 0.5 3.5V8.5C0.5 10.1569 1.84315 11.5 3.5 11.5H8.5C10.1569 11.5 11.5 10.1569 11.5 8.5V3.5C11.5 1.84315 10.1569 0.5 8.5 0.5Z" fill="" stroke="#999"></path></svg></span></label>`;
